flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1008,8 +1017,11 @@ class ExprLowering {
     resTy.push_back(converter.genType(TC, KIND));
     return genProcedureRef(funRef, resTy);
   }
+
   mlir::Value genval(const Fortran::evaluate::ProcedureRef &procRef) {
     llvm::SmallVector<mlir::Type, 1> resTy;
+    if (genAltReturnCall)
```

Comment:
It would be nice if this alternate code suggestion worked, but HasAlternateReturns() only works for a symbol that is a procedure definition.  It does not work for an external subroutine symbol that is invoked without an interface (the typical call scenario).  Worse yet, it isn't even possible to use a brute force visit of procRef's actual arguments here, because the alternate return specifiers have been eliminated from this version of the actual argument list.  It might be possible to update the front end to provide a call procRef.hasAlternateSpecifiers(),  That seems like a good solution, but I have so far been unable to figure out how to do that.

Reply:
Ok thanks; it looks like that works in some files, such as here.

Reply:
Apologies Val, I rushed and gave bad advice here... I think there is a solution to avoid the `createAltReturnCallExpression` path:

The solution lies in the arguments, after all in case of implicit interface, these are the ones that tell us what beast we are calling.

in `genProcedureRef`, if the any of the Fortran::evaluate::ActualArgument we are looping through do not contain an expression, we can check if it is an alternate return.
Below this `if`:
https://github.com/flang-compiler/f18-llvm-project/blob/5f6caa7b8b48bc9733652c15e854b3da23a4d012/flang/lib/Lower/ConvertExpr.cpp#L906
you can check that you have an alternate return argument with arg->isAlternateReturn() (see https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/include/flang/Evaluate/call.h#L111)
in which case you should set a local boolean (`genProcedureRef` scope), say `hasAlternateReturn` to `true`.
Then just before creating the function type here:
https://github.com/flang-compiler/f18-llvm-project/blob/99a79e8ab1d2df03409794297593e043bf47903c/flang/lib/Lower/ConvertExpr.cpp#L940
you could do:

```
if (hasAlternateReturn) {
 assert(resultType.size(0) && "only subroutine can have alternate retunrs");
 resultType.push_back(builder->getIndexType());
}
```

I think resultType is an ArrayRef, so you cannot push_back into it, but you should be able to create a newResultType SmallVector or such.

Then you can remove all the other changes to ConverExpr.cc (including those in `genval(const Fortran::evaluate::ProcedureRef &procRef)`)

Reply:
Jean, thanks for the suggestions, whether they turn out to be final solutions or not.  A quick take on this new suggestion is that I believe the alternate routine arguments have been removed from the `ActualArguments arguments_` list in the `ProcedurRef procRef` at this point, so the information is probably no longer recoverable.  That is of course a reasonable thing to do in some ways, because those arguments are actually nops in the call.  I'll take a more careful look at this though.  I'm also still looking for a front end fix.  Regardless, thanks for the suggestions.

Reply:
`createAltReturnCallExpression` is gone.  I have instead pursued Eric's suggestion to have the front end save the information.

Reply:
Solution looks good to me, sorry for the double bad suggestion.